### ISSUE TITLE
Yubikey support has regressed since 0.17

### ISSUE THREAD

**Issue by jdanders** [OpenSC/OpenSC#1545]
### Problem Description

I am using Arch Linux (Antergos specifically) and noticed a regression from 0.17 to 0.18. Now 0.19 is available and there is still a problem.

I use the Yubikey for both OpenVPN and pkcs15-crypt (for full disk encryption LUKS key protection). There is a regression in both usages.

In 0.19, I get the following:
```
$ opensc-tool -i
OpenSC 0.19.0 [gcc  8.2.1 20180831]
Enabled features: locking zlib readline openssl pcsc(libpcsclite.so.1)

$  pkcs15-crypt --decipher --input encrypted_key.bin --pkcs1 --raw
Using reader with a card: Yubico Yubikey NEO OTP+U2F+CCID 00 00
Enter PIN [PIN]: 
Decrypt failed: Security status not satisfied

$  openvpn myconfig.con
Sat Dec  1 18:26:52 2018 OpenVPN 2.4.6 x86_64-pc-linux-gnu [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [PKCS11] [MH/PKTINFO] [AEAD] built on Apr 24 2018
Sat Dec  1 18:26:52 2018 library versions: OpenSSL 1.1.1a  20 Nov 2018, LZO 2.10
Sat Dec  1 18:26:52 2018 PKCS#11: Adding PKCS#11 provider '/usr/lib/opensc-pkcs11.so'
NEED-OK|token-insertion-request|Please insert PIV Card Holder pin (PIV_II) token:
```
When I install OpenSC 0.17.0, both of those operations work as expected.  I'd be glad to post whatever other logging or helpful information you might need.

### Proposed Resolution
Undo whatever regressed between 0.17 and 0.19

### Steps to reproduce
See above

### Logs
https://gist.github.com/jdanders/245bfbc4cdd9c97474207fc0fc557fb6

### Other notes
Issue https://github.com/OpenSC/OpenSC/issues/1507 is still present and affects me. I built pkcs11-helper with threading disabled to work around it.

**Comment by dengert** [comment]
See https://github.com/OpenSC/OpenSC/pull/1529  It should fix this problem. It was committed 12 days ago. 
#1507  also has a fix  #1513  that is also in current master.

**Comment by frankmorgner** [comment]
> See #1529 It should fix this problem. It was committed 12 days ago.
> #1507 also has a fix #1513 that is also in current master.

Additionally, you could try Yubico's [YKCS#11](https://developers.yubico.com/yubico-piv-tool/YKCS11_release_notes.html).

**Comment by mouse07410** [comment]
While I love Yubico products, I have to advise **against** YKCS11 library, as in my experience it is functionally incomplete, buggy, and incompatible with how OpenSC numbers/names objects in the token. 

You're likely to have much better luck with this library. 

Also, do you configure "override_user_consent"? If not, I suggest you try it.

Lastly, if the cause of your problem is the software enforcement of ALWAYS_AUTHENTICATE, you can try my fork, as I disabled this feature. Reason I suggest it - I'm using Yubikey 4 and NEO for computer login (and other things that aren't relevant in this context), and the current master does fine (except the change I mentioned).

**Comment by dengert** [comment]
@mouse07410  "override_user_consent" was a suggested patch in https://github.com/OpenSC/OpenSC/issues/1533#issuecomment-440327594

https://github.com/OpenSC/OpenSC/files/2600408/pin_cache_override_user_consent_policy.diff.txt

No one appeared to be interested in it, so I never made a PR.

**Comment by dengert** [comment]
Where is the log for failing OpenVPN? 

The log above is for pkcs15-crypt and looks normal.
Yes there is a lot of activity. It is a `NEO`.  
There is a VERIFY command being sent: 
```
Outgoing APDU (13 bytes):
00 20 00 80 08 6A 6A 6A 6A 6A 6A FF FF
Incoming APDU (2 bytes):
90 00   (pin correct)
```
Which looks normal. 
Then  next APDU sequence is a  decipher using the 9C key the requires "PIN Always"  
```
Outgoing APDU (260 bytes):
10 87 07 9C FF 7C 82 01 06 82 00 81 82 01 00 45  (first line of first block with chaining flag)
Incoming APDU (2 bytes):
9000
Outgoing APDU (17 bytes):
00 87 07 9C 0B E2 3D DA 61 99 21 3F 22 2A 93 72 
00    (last part of the chained output) 
Incoming APDU (255 bytes):
7C 82 01 04 82 82 01 00 00 02 87 D8 61 B9 F9 B1 ... (First part of response)
Outgoing APDU (5 bytes):
00 C0 00 00 0B                   (get the rest of response, 0x18 bytes)
Incoming APDU (13 bytes):
4E B4 54 11 62 B6 21 E7 5D EB 5B 90 00 (rest of response, with 90 00)
```
This looks normal.

**Comment by jdanders** [comment]
Sorry, there are actually 5 logs on that gist. The one of interest is called openvpn-0.19.0.r159.gb01ca2dc-debug.log

**Comment by jdanders** [comment]
https://gist.githubusercontent.com/jdanders/245bfbc4cdd9c97474207fc0fc557fb6/raw/63ab7b3153b53d320e1b7284e4605397da2f3dba/openvpn-0.19.0.r159.gb01ca2dc-debug.log for the full file

**Comment by dengert** [comment]
```
card-piv.c:3162:piv_init: Yubico card->type=14003, r=0x00000000 version=0x00010004
card-piv.c:3221:piv_init: PIV card-type=14003 card_issues=0x00020313
```
The NEO appears to be version 1.0.4 (which I do not have), which is rather old. The card_issues that it appears to have can be found the the card-piv.c source file lines: 
```
 241 /* card_issues - bugs in PIV implementations requires special handling */
 242 #define CI_VERIFY_630X                      0x00000001U /* VERIFY tries left returns 630X rather then 63CX */
 243 #define CI_VERIFY_LC0_FAIL                  0x00000002U /* VERIFY Lc=0 never returns 90 00 if PIN not needed */
 244                                                         /* will also test after first PIN verify if protected object can be used instead */
 246 #define CI_CANT_USE_GETDATA_FOR_STATE       0x00000008U /* No object to test verification inplace of VERIFY Lc=0 */
 247 #define CI_LEAKS_FILE_NOT_FOUND             0x00000010U /* GET DATA of empty object returns 6A 82 even if PIN not verified */
 251 #define CI_OTHER_AID_LOSE_STATE             0x00000100U /* Other drivers match routines may reset our security state and lose AID!!! */
 252 #define CI_NFC_EXPOSE_TOO_MUCH              0x00000200U /* PIN, crypto and objects exposed over NFS in violation of 800-73-3 */
 255 #define CI_NO_EC384                         0x00020000U /* does not have EC 384 */
```
Not having a version of your NEO, it looks like it fixed the issue CI_VERIFY_630X as the dump 
shows `00 20 00 80` returns `63 C3`. 
Your version may still have the CI_VERIFY_LC0_FAIL but there was no pin verify command in the debug log. 
The debug log show trying to address the comment in line 244 above. It did not fine one. Thus:
`ard-piv.c:3489:piv_pin_cmd: CI_CANT_USE_GETDATA_FOR_STATE set, assume logged_in=-1`
and this was also set in the card_issues. 
`#define SC_PIN_STATE_UNKNOWN -1` But in any case the application did not provide a PIN to verify. 

`card-piv.c:3512:piv_pin_cmd: piv_pin_cmd tries_left=3, logged_in=-1`  not sure if logged in or not. 



The last line in the dump is:
`framework-pkcs15.c:587:C_GetTokenInfo: C_GetTokenInfo(0) returns 0x0`
It looks like there should be more in the debug log. 
Did the application crash? Or did you cancel it?

**Comment by jdanders** [comment]
So looking at the tail end of the two logs together, it looks like something made OpenVPN happy in 0.17 because what are the trailing lines in the 0.19 log:
```
12:09:21.056 [opensc-pkcs11] reader-pcsc.c:673:pcsc_unlock: called
12:09:21.066 [opensc-pkcs11] pkcs15-pin.c:717:sc_pkcs15_get_pin_info: returning with: 0 (Success)
12:09:21.066 [opensc-pkcs11] framework-pkcs15.c:587:C_GetTokenInfo: C_GetTokenInfo(0) returns 0x0
```
Are followed by pkcs11-global calls in the 0.17 log:
```
18:37:54.210 [opensc-pkcs11] reader-pcsc.c:662:pcsc_unlock: called
18:37:54.217 [opensc-pkcs11] pkcs15-pin.c:705:sc_pkcs15_get_pin_info: returning with: 0 (Success)
18:37:54.217 [opensc-pkcs11] framework-pkcs15.c:559:C_GetTokenInfo: C_GetTokenInfo(0) returns 0x0
18:37:54.217 [opensc-pkcs11] pkcs11-session.c:58:C_OpenSession: C_OpenSession(0x0)
18:37:54.217 [opensc-pkcs11] slot.c:448:slot_get_token: Slot(id=0x0): get token
18:37:54.217 [opensc-pkcs11] slot.c:466:slot_get_token: Slot-get-token returns OK
```
But what the difference is, I don't know. 

(BTW, Looking at the time stamps, I'm guessing I finished entering my PIN at 18:38:01.823.)

**Comment by dengert** [comment]
The best I can tell, OpenVPN has a problem. The last call to OpenSC is the C_GetTokenInfo, which returns. The question is what does OpenVPN do next, or why is it hanging. 

Might be some tread locking issue, but it does not look like it is in OpenSC. 

There is not much else I can say.

**Comment by dengert** [comment]
Yes there is. In one of the debug logs, right after the return of C_GetTokenInfo, there is a call to C_Finalize which may have been when you canceled it.

**Comment by jdanders** [comment]
I found the first issue: the pkcs11 ID has changed since 0.17. It used to be:
```
piv_II/PKCS\x2315\x20emulated/548bde42e22903b0/PIV\x20Card\x20Holder\x20pin\x20\x28PIV_II\x29/02
```
Now it is
```
piv_II/PKCS\x2315\x20emulated/548bde42e22903b0/[My CN Name]/02
```
That explains why OpenVPN didn't think there was PIV device available -- it was looking for a different ID. 

However, that doesn't solve the problem yet. Now it prompts for a PIN, but still fails to sign the request. Here is the output -- notice that it requires me to enter the PIN twice and fails on the second time.
```
Tue Dec  4 07:52:40 2018 OpenVPN 2.4.6 x86_64-pc-linux-gnu [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [PKCS11] [MH/PKTINFO] [AEAD] built on Apr 24 2018
Tue Dec  4 07:52:40 2018 library versions: OpenSSL 1.1.1a  20 Nov 2018, LZO 2.10
Tue Dec  4 07:52:40 2018 PKCS#11: Adding PKCS#11 provider '/usr/lib/opensc-pkcs11.so'
Enter [My CN Name] token Password:
Enter [My CN Name] token Password:
Tue Dec  4 07:52:52 2018 PKCS#11: Cannot perform signature 257:'CKR_USER_NOT_LOGGED_IN'
Tue Dec  4 07:52:52 2018 OpenSSL: error:141F0006:SSL routines:tls_construct_cert_verify:EVP lib
Tue Dec  4 07:52:52 2018 TLS_ERROR: BIO read tls_read_plaintext error
Tue Dec  4 07:52:52 2018 TLS Error: TLS object -> incoming plaintext read error
Tue Dec  4 07:52:52 2018 TLS Error: TLS handshake failed
```
And yes, I entered the PIN correctly :)

Here is the [opensc log](https://gist.githubusercontent.com/jdanders/245bfbc4cdd9c97474207fc0fc557fb6/raw/22abf7cd1481fa4c8dad8425cbe624dee19914c5/openvpn-0.19.0.r159.gb01ca2dc-correctID.log)

**Comment by dengert** [comment]
The PIV 9C key requires "PIN Always" and the NEO enforces it. PKCS\#15 has `user_consent` and PKCS\#11 provides the `CKA_ALWAYS_AUTHENTICATE` attribute to tell the application that it needs to do `C_Login(CKU_CONTEXT_SPECIFIC,...)`   But it looks like OpenVPN does not understand that it needs to ask for the `CKA_ALWAYS_AUTHENTICATE` attribute.  

`opensc.conf.example.in` has a way around this:
```
 885                 # Older PKCS#11 applications not supporting CKA_ALWAYS_AUTHENTICATE
 886                 # may need to set this to get signatures to work with some cards.
 887                 # Default: false
 888                 # pin_cache_ignore_user_consent = true;
 889                 #
```
The debug log shows:
`pkcs15.c:1219:sc_pkcs15_bind: PKCS#15 options: use_file_cache=0 use_pin_cache=1 pin_cache_counter=10 pin_cache_ignore_user_consent=0` and
`pkcs15-pin.c:764:sc_pkcs15_pincache_add: caching refused (user consent)` 
Make sure `pin_cache_ignore_user_consent = true` so it will try the cached pin again.

Code was added in 0.18 and 0.19 to improve detection of the applet on a card from the same or other applications.  This was especially important of Yuibico uses who wanted to use both the PIV and OpenPGP at the same time from different applications. This detection can get in the way of the "PIN Always" rule. Early NEO versions also have other issues with detection of the applet.  If the application can handle the `CKA_ALWAYS_AUTHENTICATE` the detection is handled differently to avoid interfering with "PIN Always".

**Comment by jdanders** [comment]
That seems to have fixed it. My `/etc/opensc.conf` file looks like this now:
```
app default {
	framework pkcs15 {
                pin_cache_ignore_user_consent = true
	}
}
```
Can you provide me a good description of what OpenVPN is missing so that I can open an issue in that project and get this fixed?

Now the only remaining issue is that if I use a threaded pkcs11-helper, OpenVPN locks up before getting opensc started. The OpenVPN verbose output does this:
```
Tue Dec  4 09:00:29 2018 us=351967 PKCS#11: pkcs11_initialize - entered
Tue Dec  4 09:00:29 2018 us=351988 PKCS#11: pkcs11_initialize - return 0-'CKR_OK'
Tue Dec  4 09:00:29 2018 us=351996 PKCS#11: pkcs11_addProvider - entered - provider='/usr/lib/opensc-pkcs11.so', private_mode=00000000
Tue Dec  4 09:00:29 2018 us=352015 PKCS#11: Adding PKCS#11 provider '/usr/lib/opensc-pkcs11.so'
Tue Dec  4 09:00:29 2018 us=352033 PKCS#11: pkcs11h_addProvider entry version='1.25.1', pid=2340, reference='/usr/lib/opensc-pkcs11.so', provider_location='/usr/lib/opensc-pkcs11.so', allow_protected_auth=0, mask_private_mode=00000000, cert_is_private=0
Tue Dec  4 09:00:29 2018 us=352046 PKCS#11: Adding provider '/usr/lib/opensc-pkcs11.so'-'/usr/lib/opensc-pkcs11.so'
Killed
```
The killing has to come from a `pkill -9 openvpn` otherwise it just hangs forever, ignoring Ctrl-C. This is reminiscent of Issue #1507, but I can't say for sure it's the same. 

So to be clear, bulding Git commit b01ca2d and using a threaded `pkcs11-helper`, OpenVPN hangs. If I build `pkcs11-helper` with ` --disable-threading --disable-slotevent` then OpenVPN does not hang.

**Comment by dengert** [comment]
"That patch?" do you mean https://github.com/OpenSC/OpenSC/pull/1529 ?

PKCS#15 code in OpenSC does not enforce ALWAYS_AUTHENTICATE. The PIV card does with its "PIN Always" rule. So any middleware using the card needs to do something.  The above adds code to the pkcs15-crypt.c, the only place where OpenSC provides a tool exposed to users, to tell drivers that a verify is being done that may be followed by a crypto operation that may be have some type of user_consent==1. The PIV driver is the only one that uses this currently. A PKCS\#11 `C_Login(CKU_CONTECT_SPECIFIC,...)` does the same thing. 

Minidriver may also need some work, but most people using PIV use the Microsoft minidriver for the PIV. 

 Also see  https://github.com/OpenSC/OpenSC/pull/153 where Frank andf Hanna have been discussing how user_consent should be handled.

**Comment by mouse07410** [comment]
> "That patch?" do you mean #1529 ?

No, I meant 
https://github.com/OpenSC/OpenSC/issues/1533#issuecomment-440327594. 

I think that all #1529 is doing is enforcing another login, under different name but with the same PIN (because I haven't yet seen a token that would protect SIGN key with a Pin *different* from the "global" token PIN). In my fork I disabled it. 

> The PIV card [enforces ALWAYS_AUTHENTICATE] with its "PIN Always" rule. So any middleware using the card needs to do something...

IMHO, the way to handle ALWAYS_AUTHENTICATE is compare the error to `USER_NOT_LOGGED_IN`, and arrange for re-auth if equal - *not* checking if the key is marked `ALWAYS_AUTHENTICATE` and if so, do another `VERIFY` - automatically if `ignore_user_consent` is set, or via indication to the caller that [another] login is necessary.

> Also see #153 where Frank andf Hanna have been discussing how user_consent should be handled.

Well, that was way before my time. ;-)

Regardless, I think I've heard what others think about handling of `user_consent`, and expressed my opinion on this subject.

**Comment by dengert** [comment]
Sorry I missed a digit.  #153 should have been #1539

**Comment by dengert** [comment]
@jdanders The best answer to what OpenVPN needs to support is PKCS#11 with keys with CKA_ALWAYS_AUTHENTICATE by using C_Login(CKU_CONTEXT_SPECIFIC,...)
Examples are in pkcs11-tool.c 

OpenSSH also has issues in this area. 

FireFox fixed this years ago. 

Also see: https://github.com/OpenSC/OpenSC/issues/1269

Your other choice is to (create if needed) the PIV AUTH certificate and 9A key and don't use the SIGN certificate and 9C key that requires the "PIN Always" thus avoiding this issue.

**Comment by frankmorgner** [comment]
@jdanders if the remaining issue is not fixed with #1621, please open a bug report at pkcs11-helper.

**Comment by cederom** [comment]
I have encountered similar issues with Yubikey5NFC. Using slot `9A` with the same key and certificate as slot `9C` fixed problems! It works now on FreeBSD and Windoze. Thank you guys (and @dengert for the slot hint) :-)
